README.md
Решение задачи в курсе линейной алгебры в тематике Heroes of Might and Magic III:

**Задача**:

Студент третьего курса Денис пристрастился к игре Heroes of Might and Magic III. И вот ему понадобилось набрать армию в своем городе. Следующая таблица отражает силу, стоимость и доступное для найма количество существ в городе Дениса:

| Характеристики | Титан | Нага | Джинн | Маг | Голем | Горгулья | Гремлин |
| --- | --- | --- | --- | --- | --- | --- | --- |
| Золото | 5000 | 1500 | 750 | 500 | 400 | 200 | 70 |
| Драгоценные камни | 3 | 2 | 1 | 1 | 0 | 0 | 0 |
| Ртуть | 1 | 0 | 1 | 1 | 1 | 0 | 0 |
| Сила | 300 | 120 | 60 | 40 | 35 | 20 | 4 |
| Доступное кол-во | 10 | 20 | 30 | 55 | 60 | 110 | 500 |

Денис имеет:
* 200000 золотых монет
* 115 драгоценных камней
* 80 единиц ртути

Он стремится максимизировать силу нанятой им армии, но при этом хочет, чтобы: общая сила существ дальнего боя (титанов, магов и гремлинов) была не меньше 4000 единиц, а летающих сущест (джинов и горгулий) - не меньше 2000.
Необходимо также учесть, что Денис может продать драгоценные камни на базаре по цене 500 золотых монет за один камень.

## License

Copyright © 2016 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-09-30
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-09-30
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/heroes/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/heroes/compare/0.1.0...0.1.1


test/heroes/core_test.clj
(ns heroes.core-test
  (:require [clojure.test :refer :all]
            [heroes.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/heroes/core.clj
(ns heroes.core
  (:require [clocop.core :refer :all]
            [clocop.constraints :refer :all]))

(def titan    {:gold 5000 :diamonds 3 :mercury 1 :strength 300 :available 10})
(def naga     {:gold 1000 :diamonds 2 :mercury 0 :strength 120 :available 20})
(def gin      {:gold 750  :diamonds 1 :mercury 1 :strength 60  :available 30})
(def mag      {:gold 500  :diamonds 1 :mercury 1 :strength 40  :available 55})
(def golem    {:gold 400  :diamonds 0 :mercury 1 :strength 35  :available 60})
(def gorgoyle {:gold 200  :diamonds 0 :mercury 0 :strength 20  :available 110})
(def gremlin  {:gold 70   :diamonds 0 :mercury 0 :strength 4   :available 500})

(def all-creatures [titan naga gin mag golem gorgoyle gremlin])
(def flying-creatures [gin gorgoyle])
(def range-creatures [titan mag gremlin])

(def denis-gold 200000)
(def denis-diamonds 115)
(def denis-mercury 80)

(def max-allowed-gold (+ denis-gold (* denis-diamonds 500)))

(defn -main [& args]
  (clojure.pprint/pprint
   (with-store (store)
     (let [strength       (int-var "strength"  0 (->> all-creatures
                                                      (map #(* (:strength %) (:available %)))
                                                      (apply +)))
           titans         (int-var "titans"    0 (:available titan))
           nagas          (int-var "nagas"     0 (:available naga))
           gins           (int-var "gins"      0 (:available gin))
           mags           (int-var "mags"      0 (:available mag))
           golems         (int-var "golems"    0 (:available golem))
           gorgoyles      (int-var "gorgoyles" 0 (:available gorgoyle))
           gremlins       (int-var "gremlins"  0 (:available gremlin))
           gold           (int-var "gold"      0 max-allowed-gold)
           diamonds       (int-var "diamonds"  0 denis-diamonds)
           mercury        (int-var "mercury"   0 denis-mercury)
           denis-gold     (int-var denis-gold)
           denis-diamonds (int-var denis-diamonds)
           denis-mercury  (int-var denis-mercury)
           diamond-cost   (int-var 500)]
       ;; mercury
       (constrain! ($= mercury ($+ ($* titans (:mercury titan))
                                   ($* nagas (:mercury naga))
                                   ($* gins (:mercury gin))
                                   ($* mags (:mercury mag))
                                   ($* golems (:mercury golem))
                                   ($* gorgoyles (:mercury gorgoyle))
                                   ($* gremlins (:mercury gremlin)))))
       (constrain! ($<= mercury denis-mercury))
       ;; diamonds
       (constrain! ($= diamonds ($+ ($* titans (:diamonds titan))
                                    ($* nagas (:diamonds naga))
                                    ($* gins (:diamonds gin))
                                    ($* mags (:diamonds mag))
                                    ($* golems (:diamonds golem))
                                    ($* gorgoyles (:diamonds gorgoyle))
                                    ($* gremlins (:diamonds gremlin)))))
       (constrain! ($<= diamonds denis-diamonds))
       ;; gold
       (constrain! ($= gold ($+ ($* titans (:gold titan))
                                ($* nagas (:gold naga))
                                ($* gins (:gold gin))
                                ($* mags (:gold mag))
                                ($* golems (:gold golem))
                                ($* gorgoyles (:gold gorgoyle))
                                ($* gremlins (:gold gremlin)))))
       (constrain! ($<= gold ($+ denis-gold
                                 ($* diamond-cost
                                     ($- denis-diamonds
                                         diamonds)))))
       ;; strength
       (constrain! ($>= ($+ ($* titans (:strength titan))
                            ($* mags (:strength mag))
                            ($* gremlins (:strength gremlin)))
                        4000))
       (constrain! ($>= ($+ ($* gins (:strength gin))
                            ($* gorgoyles (:strength gorgoyle)))
                        2000))
       (constrain! ($= strength ($+ ($* titans (:strength titan))
                                    ($* nagas (:strength naga))
                                    ($* gins (:strength gin))
                                    ($* mags (:strength mag))
                                    ($* golems (:strength golem))
                                    ($* gorgoyles (:strength gorgoyle))
                                    ($* gremlins (:strength gremlin)))))
       (solve! :pick-val :max)))))
